cs/src/core/Device/AzurePageBlobDevice.cs

```diff
@@ -0,0 +1,136 @@
+                {
+                    entry.Value.Delete();
+                }
+            }
+        }
+        /// <summary>
+        /// <see cref="IDevice.Close">Inherited</see>
+        /// </summary>
+        public override void DeleteSegmentRange(int fromSegment, int toSegment)
+        {
+            for (int i = fromSegment; i < toSegment; i++)
+            {
+                if (blobs.TryRemove(i, out CloudPageBlob blob))
+                {
+                    blob.Delete();
+                }
+            }
+        }
+        /// <summary>
+        /// <see cref="IDevice.ReadAsync(int, ulong, IntPtr, uint, IOCompletionCallback, IAsyncResult)">Inherited</see>
+        /// </summary>
+        public override unsafe void ReadAsync(int segmentId, ulong sourceAddress, IntPtr destinationAddress, uint readLength, IOCompletionCallback callback, IAsyncResult asyncResult)
+        {
+            CloudPageBlob pageBlob = GetOrAddPageBlob(segmentId);
+
+            // Even though Azure Page Blob does not make use of Overlapped, we populate one to conform to the callback API
+            Overlapped ov = new Overlapped(0, 0, IntPtr.Zero, asyncResult);
+            NativeOverlapped* ovNative = ov.UnsafePack(callback, IntPtr.Zero);
+
+            UnmanagedMemoryStream stream = new UnmanagedMemoryStream((byte*)destinationAddress, readLength, readLength, FileAccess.Write);
+
+            // TODO(Tianyu): This implementation seems to swallow exceptions that would otherwise be thrown from the synchronous version of this
+            // function. I wasn't able to find any good documentaiton on how exceptions are propagated or handled in this scenario. 
+            pageBlob.BeginDownloadRangeToStream(stream, (Int64)sourceAddress, readLength, ar => callback(0, readLength, ovNative), asyncResult);
+        }
+        /// <summary>
+        /// <see cref="IDevice.WriteAsync(IntPtr, int, ulong, uint, IOCompletionCallback, IAsyncResult)">Inherited</see>
+        /// </summary>
+        public override unsafe void WriteAsync(IntPtr sourceAddress, int segmentId, ulong destinationAddress, uint numBytesToWrite, IOCompletionCallback callback, IAsyncResult asyncResult)
+        {
+            CloudPageBlob pageBlob = GetOrAddPageBlob(segmentId);
+
+            // Even though Azure Page Blob does not make use of Overlapped, we populate one to conform to the callback API
+            Overlapped ov = new Overlapped(0, 0, IntPtr.Zero, asyncResult);
+            NativeOverlapped* ovNative = ov.UnsafePack(callback, IntPtr.Zero);
+            UnmanagedMemoryStream stream = new UnmanagedMemoryStream((byte*)sourceAddress, numBytesToWrite);
+            pageBlob.BeginWritePages(stream, (long)destinationAddress, null, ar => callback(0, numBytesToWrite, ovNative), asyncResult);
+        }
+
+        private CloudPageBlob GetOrAddPageBlob(int segmentId)
+        {
+            return blobs.GetOrAdd(segmentId, id => CreatePageBlob(id));
```

Comment:
>id => CreatePageBlob(id) [](start = 45, length = 24)

This factory is not guaranteed to be called exactly once - will we leak blobs in this case? Maybe use Lazy to guarantee exactly-once instantiation?